commit 839898777226a3ed88c0859f25ffe712519b4ead
Author: Rocket Ma <marocketbd@gmail.com>
Date:   Fri Apr 17 23:48:41 2026 -0700

    stdio-common: Fix buffer overflow in scanf %mc [BZ #34008]
    
    * stdio-common/vfscanf-internal.c: When enlarging allocated buffer with
    format %mc or %mC, glibc allocates one byte less, leading to
    user-controlled one byte overflow. This commit fixes BZ #34008, or
    CVE-2026-5450.
    
    Reviewed-by: Carlos O'Donell <carlos@redhat.com>
    Signed-off-by: Rocket Ma <marocketbd@gmail.com>
    Reviewed-by: H.J. Lu <hjl.tools@gmail.com>

Conflicts:
	stdio-common/Makefile
	  (fixup context)
	stdio-common/vfscanf.c
	  (vfscanf_internal is part of vfscanf.c)

diff -Nrup a/stdio-common/Makefile b/stdio-common/Makefile
--- a/stdio-common/Makefile	2026-06-02 13:10:51.282023845 -0400
+++ b/stdio-common/Makefile	2026-06-02 13:10:33.565576212 -0400
@@ -65,6 +65,7 @@ tests := tstscanf test_rdwr test-popen t
 	 tst-renameat2 \
 	 tst-printf-bz25691 \
 	 tst-vfprintf-width-prec-alloc \
+	 tst-vfscanf-bz34008 \
 	 tst-grouping2 \
   # tests
 
@@ -103,6 +104,9 @@ endif
 tst-printf-bz18872-ENV = MALLOC_TRACE=$(objpfx)tst-printf-bz18872.mtrace
 tst-vfprintf-width-prec-ENV = \
   MALLOC_TRACE=$(objpfx)tst-vfprintf-width-prec.mtrace
+tst-vfscanf-bz34008-ENV = \
+  MALLOC_CHECK_=3 \
+  LD_PRELOAD=$(common-objpfx)/malloc/libc_malloc_debug.so
 tst-printf-bz25691-ENV = \
   MALLOC_TRACE=$(objpfx)tst-printf-bz25691.mtrace

diff --git a/stdio-common/tst-vfscanf-bz34008.c b/stdio-common/tst-vfscanf-bz34008.c
new file mode 100644
index 0000000000..48371c8a3d
--- /dev/null
+++ b/stdio-common/tst-vfscanf-bz34008.c
@@ -0,0 +1,48 @@
+/* Regression test for vfscanf %Nmc out-of-bound write (BZ #34008)
+   Copyright (C) 2026 The GNU Toolchain Authors.
+   This file is part of the GNU C Library.
+
+   The GNU C Library is free software; you can redistribute it and/or
+   modify it under the terms of the GNU Lesser General Public
+   License as published by the Free Software Foundation; either
+   version 2.1 of the License, or (at your option) any later version.
+
+   The GNU C Library is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+   Lesser General Public License for more details.
+
+   You should have received a copy of the GNU Lesser General Public
+   License along with the GNU C Library; if not, see
+   <https://www.gnu.org/licenses/>.  */
+
+#include "malloc/mcheck.h"
+#include <stddef.h>
+#include <stdio.h>
+#include <string.h>
+#include <wchar.h>
+#include <stdlib.h>
+#include <malloc.h>
+#include <support/check.h>
+
+#define WIDTH 0x410
+#define SCANFSTR "%1040mc"
+static int
+do_test (void)
+{
+  mcheck_pedantic (NULL);
+  char *input = malloc (WIDTH + 1);
+  TEST_VERIFY (input != NULL);
+  memset (input, 'A', WIDTH);
+  input[WIDTH] = '\0';
+
+  char *buf = NULL;
+  TEST_VERIFY (sscanf (input, SCANFSTR, &buf) != -1);
+  TEST_VERIFY (buf != NULL);
+
+  free (buf);
+  free (input);
+  return 0;
+}
+
+#include <support/test-driver.c>
diff -Nrup a/stdio-common/vfscanf.c b/stdio-common/vfscanf.c
--- a/stdio-common/vfscanf.c	2018-08-01 01:10:47.000000000 -0400
+++ b/stdio-common/vfscanf.c	2026-06-02 13:29:00.213769470 -0400
@@ -804,8 +804,7 @@ _IO_vfscanf_internal (FILE *s, const cha
 			{
 			  /* Enlarge the buffer.  */
 			  size_t newsize
-			    = strsize
-			      + (strsize >= width ? width - 1 : strsize);
+			    = strsize + (strsize >= width ? width : strsize);
 
 			  str = (char *) realloc (*strptr, newsize);
 			  if (str == NULL)
@@ -876,7 +875,7 @@ _IO_vfscanf_internal (FILE *s, const cha
 		      && wstr == (wchar_t *) *strptr + strsize)
 		    {
 		      size_t newsize
-			= strsize + (strsize > width ? width - 1 : strsize);
+			= strsize + (strsize >= width ? width : strsize);
 		      /* Enlarge the buffer.  */
 		      wstr = (wchar_t *) realloc (*strptr,
 						  newsize * sizeof (wchar_t));
@@ -931,7 +930,7 @@ _IO_vfscanf_internal (FILE *s, const cha
 		    && wstr == (wchar_t *) *strptr + strsize)
 		  {
 		    size_t newsize
-		      = strsize + (strsize > width ? width - 1 : strsize);
+		      = strsize + (strsize >= width ? width : strsize);
 		    /* Enlarge the buffer.  */
 		    wstr = (wchar_t *) realloc (*strptr,
 						newsize * sizeof (wchar_t));
